dataset/dataset.cpp

```diff
@@ -152,14 +152,20 @@ Dataset Dataset::slice(const Slice s) const {
 }
 
 Dataset &Dataset::setSlice(const Slice s, const Dataset &data) {
-  // TODO Need dry-run mechanism here?
+  // Validate slice on all items as a dry-run
+  for (const auto &item : data)
+    (*this)[item.name()].validateSlice(s, item);
+  // Only if all items checked for dry-run does modification go-ahead
   for (const auto &item : data)
     (*this)[item.name()].setSlice(s, item);
   return *this;
```

Comment:
Wondering if it is worth saving some cost here. As it is now, in a dataset with N items, we do 2N coord comparisons, where actually only 1 is needed. To avoid this we can:

1. Call `expect::coordsAreSuperset(slice(s), data);` here (+1 coord compare).
2. Call `validateSlice` on entries of `m_items` and `data.m_items`, since these do not contain coords (-N coord compares)
3. Call `setSlice` on entries of `m_items` and `data.m_items`

Same technique can be applied below.